dao/src/main/java/org/thingsboard/server/dao/rule/BaseRuleChainService.java

```diff
@@ -758,6 +797,44 @@ public void deleteRuleNodes(TenantId tenantId, RuleChainId ruleChainId) {
         deleteEntityRelations(tenantId, ruleChainId);
     }
 
+    @Override
+    public ListenableFuture<RuleNodeStats> getRuleNodeStats(TenantId tenantId, RuleNodeId ruleNodeId) {
+        return Futures.transform(attributesService.find(tenantId, ruleNodeId, DataConstants.SERVER_SCOPE, "ruleNodeStats"), input ->
+                input.map(attributeKvEntry -> JacksonUtil.fromString(attributeKvEntry.getJsonValue().get(), RuleNodeStats.class))
+                        .orElseGet(RuleNodeStats::new), MoreExecutors.directExecutor());
+    }
+
+    private void updateRuleNodeStats(TenantId tenantId, RuleNodeId ruleNodeId, Consumer<RuleNodeStats> statsUpdater) {
+        DonAsynchron.withCallback(getRuleNodeStats(tenantId, ruleNodeId), ruleNodeStats -> {
```

Comment:
Invoking DonAsynchron.withCallback leads to the same problem as above - please use external executor service instead for db callback operations.